core/src/main/java/org/sonarsource/sonarlint/core/container/storage/StorageManager.java

```diff
@@ -0,0 +1,195 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.container.storage;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Date;
+import javax.annotation.CheckForNull;
+import org.sonarsource.sonarlint.core.client.api.GlobalConfiguration;
+import org.sonarsource.sonarlint.core.client.api.connected.GlobalSyncStatus;
+import org.sonarsource.sonarlint.core.client.api.connected.ModuleSyncStatus;
+import org.sonarsource.sonarlint.core.proto.Sonarlint;
+import org.sonarsource.sonarlint.core.util.FileUtils;
+
+public class StorageManager {
+
+  public static final String PLUGIN_REFERENCES_PB = "plugin_references.pb";
+  public static final String PROPERTIES_PB = "properties.pb";
+  public static final String MODULE_CONFIGURATION_PB = "configuration.pb";
+  public static final String RULES_PB = "rules.pb";
+  public static final String SYNC_STATUS_PB = "sync_status.pb";
+  public static final String SERVER_INFO_PB = "server_info.pb";
+  public static final String ACTIVE_RULES_FOLDER = "active_rules";
+  public static final String MODULE_LIST_PB = "module_list.pb";
+  private final Path serverStorageRoot;
+  private final Path globalStorageRoot;
+  private final Path moduleStorageRoot;
+  private final GlobalSyncStatus syncStatus;
+
+  public StorageManager(GlobalConfiguration configuration) {
+    serverStorageRoot = configuration.getStorageRoot().resolve(configuration.getServerId());
+    FileUtils.forceMkDirs(serverStorageRoot);
+    globalStorageRoot = serverStorageRoot.resolve("global");
+    FileUtils.forceMkDirs(globalStorageRoot);
+    moduleStorageRoot = serverStorageRoot.resolve("modules");
+    FileUtils.forceMkDirs(moduleStorageRoot);
+    syncStatus = initSyncStatus();
+  }
+
+  public Path getGlobalStorageRoot() {
+    return globalStorageRoot;
+  }
+
+  public Path getModuleStorageRoot(String moduleKey) {
+    return moduleStorageRoot.resolve(encodeForFs(moduleKey));
+  }
+
+  private String encodeForFs(String moduleKey) {
+    char fileSep = File.separatorChar;
+    char escape = '%';
+    int len = moduleKey.length();
+    StringBuilder sb = new StringBuilder(len);
+    for (int i = 0; i < len; i++) {
+      char ch = moduleKey.charAt(i);
+      if (ch < ' ' || ch >= 0x7F || ch == fileSep || (ch == '.' && i == 0) || ch == escape) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of conditional operators (5) used in the expression (maximum allowed 3). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1067)